example_app/src/app.rs

```diff
@@ -8,37 +11,44 @@ pub struct ExampleApp {
     frontend: Option<FrontEnd>,
 }
 
-impl epi::App for ExampleApp {
-    fn name(&self) -> &str {
-        "ewebsocket example app"
+impl ExampleApp {
+    pub fn new(cc: &eframe::CreationContext<'_>) -> Self {
+        let mut example_app = ExampleApp {
+            url: "wss://echo.websocket.events/.ws".into(),
+            ..Default::default()
+        };
+
+        example_app.connect(cc.egui_ctx.clone());
+
+        Self::default()
     }
 
-    fn setup(
-        &mut self,
-        _ctx: &egui::Context,
-        frame: &epi::Frame,
-        _storage: Option<&dyn epi::Storage>,
-    ) {
-        if let Some(web_info) = &frame.info().web_info {
-            // allow `?url=` query param
-            if let Some(url) = web_info.location.query_map.get("url") {
-                self.url = url.clone();
+    fn connect(&mut self, ctx: Context) {
+        // We can use this closure to wake up UI thread on new message
+        let wakeup = move || ctx.request_repaint();
+
+        match ewebsock::connect_with_wakeup(&self.url, wakeup) {
+            Ok((ws_sender, ws_receiver)) => {
+                self.frontend = Some(FrontEnd::new(ws_sender, ws_receiver));
+                self.error.clear();
+            }
+            Err(error) => {
+                tracing::error!("Failed to connect to {:?}: {}", &self.url, error);
+                self.error = error;
             }
         }
-        if self.url.is_empty() {
-            self.url = "wss://echo.websocket.events/.ws".into(); // echo server
-        }
-
-        self.connect(frame.clone());
     }
+}
 
-    fn update(&mut self, ctx: &egui::Context, frame: &epi::Frame) {
+impl App for ExampleApp {
+    fn update(&mut self, ctx: &egui::Context, frame: &mut eframe::Frame) {
         if !frame.is_web() {
             egui::TopBottomPanel::top("top_panel").show(ctx, |ui| {
                 egui::menu::bar(ui, |ui| {
                     ui.menu_button("File", |ui| {
                         if ui.button("Quit").clicked() {
-                            frame.quit();
+                            #[cfg(not(target_arch = "wasm32"))]
+                            frame.close();
```

Comment:
I believe this is the best way to handle the case where we're not in WASM? Maybe it would be better to just have this entire `TopBottomPanel` be scoped and `cfg`d?